File: Codes/2132-stamping-the-grid.py
'''
Author: LetMeFly
Date: 2023-12-14 19:03:32
LastEditors: LetMeFly
LastEditTime: 2023-12-14 19:25:57
'''
from typing import List

class Solution:
    def possibleToStamp(self, grid: List[List[int]], h: int, w: int) -> bool:
        n, m = len(grid), len(grid[0])
        prefix = [[0] * (m + 1) for _ in range(n + 1)]
        diff = [[0] * (m + 2) for _ in range(n + 2)]
        # get-prefix
        for i in range(n):
            for j in range(m):
                prefix[i + 1][j + 1] = grid[i][j] + prefix[i + 1][j] + prefix[i][j + 1] - prefix[i][j]
        # stamp
        for i in range(n - h + 1):
            for j in range(m - w + 1):
                # (i, j) -> (i + h - 1, j + w - 1)
                if not grid[i][j] and not (prefix[i + h][j + w] + prefix[i][j] - prefix[i + h][j] - prefix[i][j + w]):
                    diff[i + 1][j + 1] += 1
                    diff[i + h + 1][j + 1] -= 1
                    diff[i + 1][j + w + 1] -= 1
                    diff[i + h + 1][j + w + 1] += 1
        # un-diff
        for i in range(n):
            for j in range(m):
                diff[i + 1][j + 1] += diff[i + 1][j] + diff[i][j + 1] - diff[i][j]
                if not grid[i][j] and not diff[i + 1][j + 1]:
                    return False
        return True

File: Codes/2132-stamping-the-grid.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-12-14 14:43:45
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-12-14 14:49:26
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    bool possibleToStamp(vector<vector<int>>& grid, int stampHeight, int stampWidth) {
        int n = grid.size(), m = grid[0].size();
        for (int i = 0; i < n; i++) {
            int last = -1;
            for (int j = 0; j <= m; j++) {
                if (j == m || grid[i][j] == 1) {
                    if (j - last - 1 && j - last - 1 < stampWidth) {
                        return false;
                    }
                    last = j;
                }
            }
        }
        for (int j = 0; j < m; j++) {
            int last = -1;
            for (int i = 0; i <= n; i++) {
                if (i == n || grid[i][j] == 1) {
                    if (i - last - 1 && i - last - 1 < stampHeight) {
                        return false;
                    }
                    last = i;
                }
            }
        }
        return true;
    }
};

/*
[[0,0,0,0,0],[0,0,0,0,0],[0,0,1,0,0],[0,0,0,0,1],[0,0,0,1,1]] 2 2

false
0 0 0 0 0
0 0 0 0 0
0 0 1 0 0
0 0 0 0 1
0 0 0 1 1
*/